### PR TITLE
add SFSDKSafeMutableArray class

### PR DESCRIPTION
**Pull request by liguopku**
@sgoldberg-sfdc

### CODE REVIEW

**Comment by jamesward** [comment]
@guo-li I've sent you an invite to the org: https://github.com/forcedotcom

Once confirmed, refresh the CLA bot: https://cla.salesforce.com/status/forcedotcom/SalesforceMobileSDK-iOS/pull/2654

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -102,6 +102,11 @@
...
 		82D4642619C7C8170006BDFE /* CoreGraphics.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 8280EBCC16E14F7000768DE8 /* CoreGraphics.framework */; };
 		82D4642719C7C8170006BDFE /* UIKit.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 8280EBB516E14E9F00768DE8 /* UIKit.framework */; };
+		93E423A12162993D00274DF3 /* SFSDKSafeMutableArray.m in Sources */ = {isa = PBXBuildFile; fileRef = 93E423A02162993D00274DF3 /* SFSDKSafeMutableArray.m */; };
+		93E423A32162B85100274DF3 /* SFSDKSafeMutableArrayTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 93E423A22162B85100274DF3 /* SFSDKSafeMutableArrayTests.m */; };
+		93E423A42162DAF200274DF3 /* SFSDKSafeMutableArray.h in Headers */ = {isa = PBXBuildFile; fileRef = 93E4238C216297BA00274DF3 /* SFSDKSafeMutableArray.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		93E423A52162DB0E00274DF3 /* SFSDKSafeMutableArray.h in Headers */ = {isa = PBXBuildFile; fileRef = 93E4238C216297BA00274DF3 /* SFSDKSafeMutableArray.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (trooper2013) [inline libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: Duplicate entry here?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    dispatch_sync(self.queue, ^{
+        array = [NSArray arrayWithArray:self.backingArray];
+    });
+    return array;
+}
+
+- (void)filterUsingPredicate:(NSPredicate *)predicate {
+    dispatch_sync(self.queue, ^{
+        [self.backingArray filterUsingPredicate:predicate];
+    });
```
Comment (sgoldberg-sfdc) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m]: Doesnt this mutate the backing array? I think it should use the barrier.

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (bhariharan) [reply libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: Make sure it's added to both targets (static and dynamic framework).

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m (hunk 2 above)
Comment (liguopku) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m]: oh let me update it.

File: libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (liguopku) [reply libs/SalesforceSDKCore/SalesforceSDKCore.xcodeproj/project.pbxproj]: I added the file to both targets. 
screen shot 2018-10-02 at 10 32 03 am

**Comment by liguopku** [comment]
@trooper2013 I checked the duplicated entries. they are required for the two targets in the libs. @sgoldberg-sfdc I update to use barrir in filter method.

**Comment by liguopku** [comment]
@jamesward I accept the invitation but still see cla verification failed...could you help to take a look if I missed anything?

**Comment by jamesward** [comment]
@liguopku The user the associated with the commits is @guo-li and I just invited that user to the org.  Is that another account you use?

**Comment by liguopku** [comment]
ohh I see. Guo-Li is the enterprise user I use. Liguopku is my public github user.

**Comment by jamesward** [comment]
Those commits are associated with the https://github.com/guo-li user.  So you can either change the commit authors (https://help.github.com/articles/changing-author-info/) or accept the invite as the guo-li user.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+}
+
++ (id)arrayWithCapacity:(NSUInteger)numItems {
+    id retVal = [[self alloc] initWithCapacity:numItems];
+    return retVal;
+}
+
+#pragma private methods
+- (void)initQueue {
+    self.queue = dispatch_queue_create([NSString stringWithFormat:@"com.salesforce.mobilesdk.readWriteSetQ%u", arc4random_uniform(UINT32_MAX)].UTF8String, DISPATCH_QUEUE_CONCURRENT);
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Common/SFSDKSafeMutableArray.m]: Replace readWriteSetQ with readWriteArrayQ

**Pull request merged**